tests/core/pyspec/eth2spec/test/helpers/hive.py

```diff
@@ -0,0 +1,389 @@
+        }
+        return self
+
+    def next_sync_committee(self, next_sync_committee):
+        self.fields["next_sync_committee"] = {
+            "pubkeys": [pubkey.hex() for pubkey in next_sync_committee.pubkeys],
+            "aggregate_pubkey": next_sync_committee.aggregate_pubkey.hex(),
+        }
+        return self
+
+    def from_state(self, state):
+        """
+        Constructs a full state verification object from the given ``state``.
+        Might be too expensive because it checks every single field, and the
+        produced yaml file is too big.
+        """
+        # Common Attributes
+        self.state_root(state.hash_tree_root())
+        self.slot(state.slot)
+        self.genesis_time(state.genesis_time)
+        self.genesis_validators_root(state.genesis_validators_root)
+        self.fork(state.fork)
+        self.latest_block_header(state.latest_block_header)
+        self.block_roots(state.block_roots)
+        self.state_roots(state.state_roots)
+        self.historical_roots(state.historical_roots)
+
+        # Eth1
+        self.eth1_data(state.eth1_data)
+        self.eth1_data_votes(state.eth1_data_votes)
+        self.eth1_deposit_index(state.eth1_deposit_index)
+
+        # Registry
+        self.validators(state.validators)
+        self.balances(state.balances)
+
+        # Randomness
+        self.randao_mixes(state.randao_mixes)
+
+        # Slashings
+        self.slashings(state.slashings)
+
+        # Attestations / Participation
+        if hasattr(state, "previous_epoch_attestations"):
+            self.previous_epoch_attestations(state.previous_epoch_attestations)
+        else:
+            self.previous_epoch_participation(state.previous_epoch_participation)
+        if hasattr(state, "current_epoch_attestations"):
+            self.current_epoch_attestations(state.current_epoch_attestations)
+        else:
+            self.current_epoch_participation(state.current_epoch_participation)
+
+        # Finality
+        # TODO: self.justification_bits(state.justification_bits)
+        self.previous_justified_checkpoint(state.previous_justified_checkpoint)
+        self.current_justified_checkpoint(state.current_justified_checkpoint)
+        self.finalized_checkpoint(state.finalized_checkpoint)
+
+        # Altair
+        if hasattr(state, "inactivity_scores"):
+            self.inactivity_scores(state.inactivity_scores)
+        if hasattr(state, "current_sync_committee"):
+            self.current_sync_committee(state.current_sync_committee)
+        if hasattr(state, "next_sync_committee"):
+            self.next_sync_committee(state.next_sync_committee)
+
+        return self
+
+        # TODO: Bellatrix, Capella, Deneb fields
+
+
+def CheckpointToDict(checkpoint):
+    return {
+        "epoch": int(checkpoint.epoch),
+        "root": checkpoint.root.hex(),
+    }
+
+
+@dataclass
+class EthV1BeaconStatesFinalityCheckpoints:
```

Comment:
So do you create a data-class per beacon-api that you test?

How does this show up in the yaml outputs? Can you share a sample output?

Reply:
Exactly, yes, it's basically a list of actions/verifications in YAML format, and the hive simulator is basically just a parser that sets up the client and follows each step from this list one by one:
```yaml
- !EthV1BeaconStatesFinalityCheckpoints
  id: head
  finalized: false
  execution_optimistic: null
  data:
    previous_justified: {epoch: 0, root: '0000000000000000000000000000000000000000000000000000000000000000'}
    current_justified: {epoch: 0, root: '0000000000000000000000000000000000000000000000000000000000000000'}
    finalized: {epoch: 0, root: '0000000000000000000000000000000000000000000000000000000000000000'}
- !EthV1BeaconStatesFork
  id: head
  finalized: false
  execution_optimistic: null
  data: {previous_version: 0200000a, current_version: 0300000a, epoch: 0}
```
On hive's end, these look like this: https://github.com/ethereum/hive/blob/572390c1559196597a8bbdf6c1e404c34d2caad8/simulators/beacon/api/verifications.go#L520-L561

My plan was to add not only verifications but also actions like send an specific message through P2P and expect a certain output to verify.